### PR TITLE
VCMI: modernize lib/serializer and modernize lib/spells

### PR DESCRIPTION
**Pull request by rilian-la-te**
More modernizations. Splitted to be easy for review

### CODE REVIEW

File: lib/serializer/BinaryDeserializer.cpp
Comment (IvanSavenko) [inline lib/serializer/BinaryDeserializer.cpp:27]: What side effects? Please describe reason in comment in code

File: lib/serializer/BinarySerializer.cpp
Comment (IvanSavenko) [inline lib/serializer/BinarySerializer.cpp:27]: Same here

File: lib/serializer/CMemorySerializer.h
Comment (IvanSavenko) [inline lib/serializer/CMemorySerializer.h:23]: Probably better to put this in constructor, to have initialization of all members in a single location.

File: lib/serializer/BinaryDeserializer.cpp
Comment (IvanSavenko) [inline lib/serializer/BinaryDeserializer.cpp:64]: I thought that clang-tidy would replace c-style casts with static_cast/reinterpret_cast?

File: lib/serializer/CTypeList.h
Comment (IvanSavenko) [inline lib/serializer/CTypeList.h:88]: Again, don't see a reason to declare method static unless method *needs* to be called without isntance of object.
Please 
a) convert this into a static function insider .cpp file
b) revert change (probably better IMO)

File: lib/serializer/Connection.h
Comment (IvanSavenko) [inline lib/serializer/Connection.h:82]: Please, keep initialization of all members in one place, either in constructor or in-place.
Don't want to play "find where this field is initialized" game.

File: lib/serializer/JsonDeserializer.h
Comment (IvanSavenko) [inline lib/serializer/JsonDeserializer.h:40]: Same here. I think better to remove this "modernizing" rule entirely

File: lib/spells/CSpellHandler.cpp
Comment (IvanSavenko) [inline lib/spells/CSpellHandler.cpp:93]: Same

File: lib/spells/CSpellHandler.h
Comment (IvanSavenko) [inline lib/spells/CSpellHandler.h:198]: I think better to initialize all of these in constructor.
1) This creates additional "noise" if you simply want to check public interface of a class
2) This also requires providing access to enum values in header files. Meaning, additional includes in header that can't be removed, slowing down compilation, especially - rebuilds where you have to change some common header

IMO in-place initialization is nice for small C-like structs, but for classes better to use constructor init

File: lib/spells/TargetCondition.cpp
Comment (IvanSavenko) [inline lib/spells/TargetCondition.cpp:268]: Please keep initializers in one place. It is weird to have selector initialized in constructor and cachingString in-place

File: lib/serializer/BinaryDeserializer.cpp
Comment (IvanSavenko) [inline lib/serializer/BinaryDeserializer.cpp:100]: Change of behavior - now `read()` will overwrite `text` since they both point to same variable.
Is this clang-tidy change or manual one?

**Review by IvanSavenko — COMMENTED** [review]
On making methods static:
I think we should remove this rule. 
Yes, this means that function will get `this` as parameters, which, oh no, will use 8 bytes on stack and several nanoseconds of CPU time if compiler fails at optimizing.

But IMO static methods should be reserved to methods that *should* be called without an instance.

File: lib/serializer/BinaryDeserializer.cpp
Comment (rilian-la-te) [reply lib/serializer/BinaryDeserializer.cpp:64]: Not all, but I will try with version 15 too.

File: lib/serializer/BinaryDeserializer.cpp
Comment (rilian-la-te) [reply lib/serializer/BinaryDeserializer.cpp:100]: clang-tidy one.

File: lib/serializer/BinaryDeserializer.cpp
Comment (rilian-la-te) [reply lib/serializer/BinaryDeserializer.cpp:27]: Side effects is always same - something stops compiling or linking, because destructor implicitly requires knowledge about something included only in *.cpp. I need to explain this in each destructor?

File: lib/serializer/CMemorySerializer.h
Comment (rilian-la-te) [reply lib/serializer/CMemorySerializer.h:23]: @IvanSavenko I can manually try put initializations into a constructor when constructor preserved, but in many cases moving init from a constructor to class can lead to removal of entire constructor.

File: lib/serializer/BinaryDeserializer.cpp
Comment (IvanSavenko) [reply lib/serializer/BinaryDeserializer.cpp:27]: `//must be instantiated in .cpp file for access to complete types of all member fields`

Comment (IvanSavenko) [reply lib/serializer/BinaryDeserializer.cpp:27]: Problem is "side effect" is too generic. Could be some weird optimization issue on some platform, could be even compiler bug.

File: lib/serializer/CMemorySerializer.h
Comment (IvanSavenko) [reply lib/serializer/CMemorySerializer.h:23]: Perhaps we can in general remove this rule and only use it manually for small structs.
For example, in this PR `ObjectPosInfo` is the only object where in-place initialization is clearly better.

I don't think that we have large number of "dump structs", with netpacks as possible exception. And for netpacks, we can try processing it with clang-tidy separately.

File: lib/serializer/BinaryDeserializer.cpp
Comment (IvanSavenko) [reply lib/serializer/BinaryDeserializer.cpp:64]: If I am reading this correctly, looks to be present, including auto-fix. And available in older versions.
https://releases.llvm.org/12.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/cppcoreguidelines-pro-type-cstyle-cast.html

Comment (rilian-la-te) [reply lib/serializer/BinaryDeserializer.cpp:64]: Yes, it is present. But this rule is present too: ```This rule bans (T)expression only when used to perform an unsafe cast.``` Maybe clang-tidy thinks than cast is safe here.

File: lib/spells/CSpellHandler.h
Comment (rilian-la-te) [reply lib/spells/CSpellHandler.h:198]: @IvanSavenko is constructor removed here? If this struct becomes trivial - then I think we should initialize all fields in class there. If not - we can again create default constructor with initialized fields.

File: lib/serializer/BinaryDeserializer.cpp
Comment (IvanSavenko) [reply lib/serializer/BinaryDeserializer.cpp:64]: It might be safe here, but here it is definitely NOT safe:
https://github.com/vcmi/vcmi/pull/1561#discussion_r1100218176

Well, I guess we should look into that warning eventually...

**Comment by rilian-la-te** [comment]
Where now-static town portal methods should be located? In CHeroInstance, or maybe in CGameState?

Because it does not needs logically any spell mechanics - it need only position and town list (and in second case - also Player Team and Player State).

**Comment by IvanSavenko** [comment]
Another interesting option is to have Castle Gate execute TownPortalMechanics with custom parameters.

I'd say that what needs refactoring is town buildings in general. Town portal logic looks OK, problem is with town building. And Castle Gate is not the only such building.

Probably better to leave Castle Gate/ Town Portal as it for now, and change only after there is a clear idea as to how refactor town buildings support.

**Comment by rilian-la-te** [comment]
@IvanSavenko I think than we should have some "Action" than can be applied both to adventure spells and town buildings. For example, TRADE action can be applied both to Marketplace and to hypothetical ALCHEMY spell, and to market adventure map object. Also TO_TOWN action with Town Portal and Castle Gate, Visions action to thieves and Vision spell, and so on. If bonuses can be actionable, than it is a good place to have it use bonuses instead.

File: lib/serializer/CMemorySerializer.h
Comment (rilian-la-te) [reply lib/serializer/CMemorySerializer.h]: done

File: lib/serializer/BinarySerializer.cpp
Comment (rilian-la-te) [reply lib/serializer/BinarySerializer.cpp]: done

File: lib/serializer/BinaryDeserializer.cpp
Comment (rilian-la-te) [reply lib/serializer/BinaryDeserializer.cpp]: reverted

File: lib/serializer/CTypeList.h
Comment (rilian-la-te) [reply lib/serializer/CTypeList.h]: reverted

File: lib/spells/CSpellHandler.cpp
Comment (rilian-la-te) [reply lib/spells/CSpellHandler.cpp]: done

File: lib/spells/CSpellHandler.h
Comment (rilian-la-te) [reply lib/spells/CSpellHandler.h]: moved to constructor again

File: lib/spells/TargetCondition.cpp
Comment (rilian-la-te) [reply lib/spells/TargetCondition.cpp:259]: all to fields done

File: lib/serializer/BinaryDeserializer.cpp
Comment (rilian-la-te) [reply lib/serializer/BinaryDeserializer.cpp]: reverted

File: lib/serializer/JsonDeserializer.h
Comment (rilian-la-te) [reply lib/serializer/JsonDeserializer.h:40]: About LIC - it needed to move out of classes, because it is just duplicated as is.

Comment (rilian-la-te) [reply lib/serializer/JsonDeserializer.h]: Done as new interface.

**Comment by IvanSavenko** [comment]
By the way, found `-Wold-style-cast` in clang-tidy: `google-readability-casting` that "suggests automated fixes in some cases". Might be worth trying it.
It will also affect casts between arithmetic types, but that's OK IMO.

File: lib/serializer/JsonUpdater.cpp
Comment (IvanSavenko) [inline lib/serializer/JsonUpdater.cpp:70]: I would rather un-comment this.
"unused variable" warning is disabled anyway

File: lib/spells/CSpellHandler.h
Comment (IvanSavenko) [inline lib/spells/CSpellHandler.h:328]: Another conversion to static

File: lib/serializer/JsonSerializeFormat.cpp
Comment (rilian-la-te) [reply lib/serializer/JsonSerializeFormat.cpp:109]: @IvanSavenko Will convert it, it was in first pass, but I disabled this check on second pass.

File: lib/serializer/JsonUpdater.cpp
Comment (rilian-la-te) [reply lib/serializer/JsonUpdater.cpp]: done

File: lib/spells/CSpellHandler.h
Comment (rilian-la-te) [reply lib/spells/CSpellHandler.h]: Done

File: lib/spells/ISpellMechanics.cpp
Comment (rilian-la-te) [reply lib/spells/ISpellMechanics.cpp]: Done

File: lib/serializer/Connection.h
Comment (IvanSavenko) [reply lib/serializer/Connection.h:82]: @rilian-la-te I think you missed this comment?

Comment (rilian-la-te) [reply lib/serializer/Connection.h]: done. This was clang-tidy bogus initialization.

**Review by IvanSavenko — APPROVED** [review]
Looks OK.
Anyone else who want to check this before merge? @kambala-decapitator ?

File: lib/serializer/JsonSerializeFormat.cpp
Comment (rilian-la-te) [reply lib/serializer/JsonSerializeFormat.cpp:109]: @IvanSavenko added modernize-pass-by-value again, will try to build VCMI with fixes to pass-by-value.

**Comment by IvanSavenko** [comment]
By the way, @rilian-la-te can you show your current clang-tidy settings that were used for this PR?

**Comment by rilian-la-te** [comment]
@IvanSavenko @kambala-decapitator up-to-date clang-tidy settings is in https://github.com/rilian-la-te/vcmi/tree/clang-tidy-v2, and the rest of modernized lib code is there too.

**Comment by IvanSavenko** [comment]
@rilian-la-te on clang-tiny config:
I compared yours to the one I setup for myself (but for IDE highlight and not for auto-conversion).
I think you can:
1) restore currently disabled `-modernize-avoid-bind`
2) disable `readability-convert-member-functions-to-static` if you haven't already.

There are few other checks which are not really useful for us, but they also don't offer auto-conversion so should be OK for your purposes

Also, FYI - clang-tidy config can be used with more readable format, without 500+ symbols in a single line, e.g:
```
Checks: >
  -*,
  modernize-*,
  -modernize-use-trailing-return-type,
  -modernize-return-braced-init-list

```

**Comment by rilian-la-te** [comment]
@IvanSavenko modernize-avoid-bind breaking spells, I do not know how to rewrite it. About static - I remained it enabled and then edit by hand, because it can show us really corner cases like ILICReader)

**Comment by IvanSavenko** [comment]
Understood. Well, I think both rules can be postponed for now then.
Besides, reviewing changes from a single rule might be better than review of multiple rules in one subsystem.

**Comment by rilian-la-te** [comment]
@IvanSavenko generate changes by one rule per line is more difficult than generate changes at once and then post subsystem by subsystem.

**Comment by rilian-la-te** [comment]
@kambala-decapitator waiting for your verdict and merge)

File: lib/serializer/ILICReader.cpp
Comment (kambala-decapitator) [inline lib/serializer/ILICReader.cpp:49]: please insert empty line at EOF

File: lib/serializer/ILICReader.h
Comment (kambala-decapitator) [inline lib/serializer/ILICReader.h:19]: why `const bool` param?

File: lib/serializer/CTypeList.h
Comment (kambala-decapitator) [inline lib/serializer/CTypeList.h:114]: which "above" does it refer to?

File: lib/serializer/CSerializer.h
Comment (kambala-decapitator) [inline lib/serializer/CSerializer.h:88]: so what's our stance on declaring default ctor/dtor explicitly? In the PR I see a mix of approaches.

File: lib/serializer/CMemorySerializer.cpp
Comment (kambala-decapitator) [inline lib/serializer/CMemorySerializer.cpp:35]: `readPos` value could be moved to declaration I guess

File: lib/serializer/JsonDeserializer.h
Comment (kambala-decapitator) [inline lib/serializer/JsonDeserializer.h:13]: please remove double space

File: lib/serializer/JsonUpdater.h
Comment (kambala-decapitator) [inline lib/serializer/JsonUpdater.h:13]: please remove double space

File: lib/spells/BattleSpellMechanics.cpp
Comment (kambala-decapitator) [inline lib/spells/BattleSpellMechanics.cpp:456]: that's a very strange mix of spacing... Any chance to unify? Per current code style `&` and `*` should actually be surrounded by spaces.

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: lib/serializer/ILICReader.h
Comment (rilian-la-te) [reply lib/serializer/ILICReader.h:19]: @kambala-decapitator because it was in both JsonUpdater.cpp + JsonSerializer.cpp. I can remove const if desired.

File: lib/serializer/CSerializer.h
Comment (rilian-la-te) [reply lib/serializer/CSerializer.h:88]: @kambala-decapitator we cannot skip declaration here, because then it will fail to link.

File: lib/serializer/CMemorySerializer.cpp
Comment (rilian-la-te) [reply lib/serializer/CMemorySerializer.cpp:35]: It can, but as discussed in Slack, @IvanSavenko tells me than he prefers this approach - if we have a non-default constructor, initalize all fields in constructor, not in declaration.

What is better?

File: lib/serializer/ILICReader.cpp
Comment (rilian-la-te) [reply lib/serializer/ILICReader.cpp]: done

File: lib/serializer/JsonUpdater.h
Comment (rilian-la-te) [reply lib/serializer/JsonUpdater.h]: done

File: lib/serializer/CTypeList.h
Comment (rilian-la-te) [reply lib/serializer/CTypeList.h:114]: It seems to be lost in ages. But this operator will always assert, so, I think deleting it instead is better, because it just not compile if requested instead of runtime error.

File: lib/serializer/ILICReader.h
Comment (rilian-la-te) [reply lib/serializer/ILICReader.h]: @kambala-decapitator `const bool` parameter changed to `bool`. But it is still `const bool` semantically, because it should not be changed.

File: lib/spells/BattleSpellMechanics.cpp
Comment (rilian-la-te) [reply lib/spells/BattleSpellMechanics.cpp]: passed via clang-format all.

File: lib/serializer/JsonDeserializer.h
Comment (rilian-la-te) [reply lib/serializer/JsonDeserializer.h]: done

File: lib/serializer/CTypeList.h
Comment (kambala-decapitator) [reply lib/serializer/CTypeList.h:114]: thanks! deleting is fine, I just couldn't understand what this comment means.

File: lib/serializer/CSerializer.h
Comment (kambala-decapitator) [reply lib/serializer/CSerializer.h:88]: I don't think I understand... If default constructor is removed, compiler would synthesise one automatically. How could that lead to a link error? Or is there another ctor in the class?

File: lib/serializer/CMemorySerializer.cpp
Comment (kambala-decapitator) [reply lib/serializer/CMemorySerializer.cpp:35]: aha, let's keep then

File: lib/serializer/ILICReader.h
Comment (kambala-decapitator) [reply lib/serializer/ILICReader.h]: at the calling site it wouldn't change anyway. tbh I've never seen declaring POD type params as `const`.

File: lib/serializer/CSerializer.h
Comment (IvanSavenko) [reply lib/serializer/CSerializer.h:88]: I think we should drop explicitly defaulted constructor/destructor if possible.

In this particular case it is probably not necessary?
Generally, in order for compiler to generate destructor compiler needs access to all class members (NOT just forward-declaration).
Because of this, if (for example) your class have shared_ptr<XXX> as member, XXX can be forward declared in header. However to emit destructor, compiler will need access to actual declaration of XXX. Meaning, destructor for such class must be defined in a place that has XXX declaration visible. Normally, cpp file is such place.

I think same applies for constructor, but never encounter this myself before.

Also, I won't be surprised if there are few other cases related to dll linkage.

File: lib/serializer/CMemorySerializer.cpp
Comment (IvanSavenko) [reply lib/serializer/CMemorySerializer.cpp:35]: I would rather keep it here. 
Not 100% sure what solution in regards to in-place initialization is better, so I think for now we should avoid too many changes. Let's do parts on which everyone agrees. That's more than enough things to do anyway.

File: lib/serializer/CSerializer.h
Comment (rilian-la-te) [reply lib/serializer/CSerializer.h:88]: @kambala-decapitator @IvanSavenko I am rechecking this now, if link failure will not remain after clang and gcc upgrade, there can be a compiler bug or my build bug (somewhere I can forgot to do clean-build).

Comment (rilian-la-te) [reply lib/serializer/CSerializer.h]: After upgrade it seems to linking normally without a constructor after clean-build.
If CI passes - I think constructor here can safely be removed. Changes pushed.

Comment (rilian-la-te) [reply lib/serializer/CSerializer.h]: CI passed, so, constructor removal is fine.

Comment (kambala-decapitator) [reply lib/serializer/CSerializer.h]: the PR still has a mix of approaches. Can default ctors/dtors be removed from everywhere?

**Comment by kambala-decapitator** [comment]
seems fine now except the last point about default ctors/dtors

File: lib/serializer/CSerializer.h
Comment (rilian-la-te) [reply lib/serializer/CSerializer.h]: @kambala-decapitator dtors - certainly not, there is a places when dtor should be explicitly set to default in *.cpp file.

Comment (rilian-la-te) [reply lib/serializer/CSerializer.h]: And also - when we need to annotate than destructor is virtual. This is done non-consistent now, but should be more consistent. 

About ctors - I thought I removed all defaulted ctors.

Comment (IvanSavenko) [reply lib/serializer/CSerializer.h]: Yeah, looks to be mostly consistent now.
Searching for word 'default' in PR I see following scenario:
1) Constructor defined in-place as default: required, since class has 2nd constructor, so compiler will normally delete default (0-params) constructor.
2) Destructor defined as default in header: required, since this is virtual destructor of a base class
3) Destructor defined as default in source file: required for compiler to generate destructors for members.

I see only one inconsistency: sometimes destructor is declared in header and sometimes in source. (case 2 & 3 from above). We *can* move all of them in .cpp, but not sure if this is better style-wise. Any opinions, @kambala-decapitator ?

Personally fine either way IMO.

Comment (IvanSavenko) [reply lib/serializer/CSerializer.h]: > And also - when we need to annotate than destructor is virtual

Well, those can be marked as `override`, (except for base class obviously). Already saw that in other PR from you. Looks OK I think.

Comment (rilian-la-te) [reply lib/serializer/CSerializer.h]: @IvanSavenko I mean your case 2) If I have not removed virtual or override destructor in derived class - it is my fault.

Comment (kambala-decapitator) [reply lib/serializer/CSerializer.h]: declaring virtual dtor is clear.

> dtors - certainly not, there is a places when dtor should be explicitly set to default in *.cpp file.

I don't think I get which case is that (also case number 3)... Could you show an example? Here's an example with forward declaration which I thought was related to this, but it works fine either if you place default dtor in header or remove the dtor from code, in both cases the `die` message is printed:

```
// 1.h
struct A {
	~A();
};
```

```
// 1.cpp
#include "1.h"
#include <iostream>
A::~A() { std::cout << "die\n"; }
```

```
// 2.h
struct A;
struct B {
	A& a;
	
	B(A& a_);
	~B() = default;
};
```

```
#include "1.h"
#include "2.h"
B::B(A& a_) : a{a_} {}

int main()
{
	A a;
	B b{a};
	return 0;
}
```

Comment (kambala-decapitator) [reply lib/serializer/CSerializer.h]: even if I move `main()` to another file and remove `#include "1.h"` from `2.cpp` there's still no compile/link error

Comment (rilian-la-te) [reply lib/serializer/CSerializer.h]: @kambala-decapitator to get a link error, you need to slightly adjust an example in file 2.h

```c++
// 2.h
struct A;
struct B {
	std::unique_ptr<A> a;
	
	B(A& a_);
	~B() = default;
};
```

Then it will be link error, because to destroy unique_ptr, you need full signature of A.

Comment (kambala-decapitator) [reply lib/serializer/CSerializer.h]: now it's fully clear, thank you very much!

**Review by kambala-decapitator — APPROVED** [review]

**Pull request merged**
